src/Lib.P2P/Multiplex/Muxer.cs

```diff
@@ -0,0 +1,340 @@
+        {
+            if (Substreams.TryRemove(stream.Id, out _))
+
+                // Tell the other side.
+                using (await AcquireWriteAccessAsync().ConfigureAwait(false))
+                {
+                    var header = new Header
+                    {
+                        StreamId = stream.Id,
+                        PacketType = PacketType.CloseInitiator
+                    };
+                    await header.WriteAsync(Channel, cancel).ConfigureAwait(false);
+                    Channel.WriteByte(0); // length
+                    await Channel.FlushAsync().ConfigureAwait(false);
+                }
+
+            return stream;
+        }
+
+        /// <summary>
+        ///   Read the multiplex packets.
+        /// </summary>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        /// <remarks>
+        ///   A background task that reads and processes the multiplex packets while
+        ///   the <see cref="Channel"/> is open and not <paramref name="cancel">cancelled</paramref>.
+        ///   <para>
+        ///   Any encountered errors will close the <see cref="Channel"/>.
+        ///   </para>
+        /// </remarks>
+        public async Task ProcessRequestsAsync(CancellationToken cancel = default(CancellationToken))
+        {
+            try
+            {
+                while (Channel.CanRead && !cancel.IsCancellationRequested)
+                {
+                    // Read the packet prefix.
+                    var header = await Header.ReadAsync(Channel, cancel).ConfigureAwait(false);
+                    var length = await Varint.ReadVarint32Async(Channel, cancel).ConfigureAwait(false);
+                    if (log.IsTraceEnabled)
+                        log.TraceFormat("received '{0}', stream={1}, length={2}", header.PacketType, header.StreamId,
+                            length);
+
+                    // Read the payload.
+                    var payload = new byte[length];
+                    await Channel.ReadExactAsync(payload, 0, length, cancel).ConfigureAwait(false);
+
+                    // Process the packet
+                    Substreams.TryGetValue(header.StreamId, out var substream);
+                    switch (header.PacketType)
+                    {
+                        case PacketType.NewStream:
+                            if (substream != null)
+                            {
+                                log.Warn($"Stream {substream.Id} already exists");
+                                continue;
+                            }
+
+                            substream = new Substream
+                            {
+                                Id = header.StreamId,
+                                Name = Encoding.UTF8.GetString(payload),
+                                Muxer = this
+                            };
+                            if (!Substreams.TryAdd(substream.Id, substream))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)